src/Compilers/Core/Portable/PEWriter/MetadataHeapsBuilder.cs

```diff
@@ -305,24 +305,38 @@ private void SerializeStringHeap()
             foreach (KeyValuePair<string, StringIdx> cur in sorted)
             {
                 uint position = _stringWriter.BaseStream.Position + (uint)_stringIndexStartOffset;
+                var str = cur.Key;
 
                 // It is important to use ordinal comparison otherwise we'll use the current culture!
-                if (prev.EndsWith(cur.Key, StringComparison.Ordinal))
+                if (prev.EndsWith(str, StringComparison.Ordinal))
                 {
                     // Map over the tail of prev string. Watch for null-terminator of prev string.
-                    _stringIndexMap[cur.Value.VirtIdx] = position - (uint)(s_utf8Encoding.GetByteCount(cur.Key) + 1);
+                    _stringIndexMap[cur.Value.VirtIdx] = position - (uint)(s_utf8Encoding.GetByteCount(str) + 1);
                 }
                 else
                 {
                     _stringIndexMap[cur.Value.VirtIdx] = position;
 
-                    // TODO (tomat): consider reusing the buffer instead of allocating a new one for each string
-                    _stringWriter.WriteBytes(s_utf8Encoding.GetBytes(cur.Key));
+                    // PERF: Save byte[] allocations by doing our own UTF8 conversion until we encounter a non-ASCII character.
+                    for (int i = 0; i < str.Length; i++)
+                    {
+                        char ch = str[i];
+                        if (ch >= 0x80)
+                        {
+                            byte[] bytes = new byte[s_utf8Encoding.GetByteCount(str) - i];
```

Comment:
Look a little closer, @gafter.
This code path is hit only for non-ASCII chars, and it's hit only once, the first time we encounter such a char (there's a 'break' after the WriteBytes) to handle the rest of the string.

In fact, worst case (when the first char is non-ASCII), this code reduces to the original code because the original "GetBytes(string s)" call is implemented in the same way: A call to GetByteCount to allocate a byte[], then a call to GetBytes to fill the array.